### PR TITLE
feat: gated sparse autoencoders

### LINKED ISSUE #103: [Proposal] Implement gated SAEs
### Proposal 

Implemented gated SAEs as described here: 
https://arxiv.org/abs/2404.16014

### Motivation

Gated SAEs are reported to be strictly better than regular SAEs

### Pitch

Implement a new `GatedSparseAutoencoder` class, or whatever way is best

### Alternatives

A clear and concise description of any alternative solutions or features you've considered, if any.

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/jbloomAus/SAELens/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by dtch1997**
# Description

Implements gated sparse autoencoders, which were recently found to solve the shrinkage problem and achieve equal reconstruction loss with 2x average feature sparsity 

Fixes #103 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### You have tested formatting, typing and unit tests (acceptance tests not currently in use)

- [x] I have run `make check-ci` to check format and linting. (you can run `make format` to format code if needed.)

### Performance Check.

If you have implemented a training change, please indicate precisely how performance changes with respect to the following metrics:
- [ ] L0
- [ ] CE Loss
- [ ] MSE Loss
- [ ] Feature Dashboard Interpretability

Please links to wandb dashboards with a control and test group.

### CODE REVIEW

**Comment by dtch1997** [comment]
Wandb run ongoing: https://wandb.ai/dtch1997/gpt2?nw=nwuserdtch1997

**Comment by dtch1997** [comment]
Cluster is being annoying. Here's a colab to run the training. 
https://colab.research.google.com/drive/1q_X46e0eO9Q1_t283USX1QG41Q5fgDAZ?usp=sharing

**Comment by dtch1997** [comment]
Comment from Neel: 

> Awesome! I asked Sen (the lead author) for thoughts:
The sparsity penalty isn't comparable between the two setups (or even the relationship between L1 and L0 norms IIRC). So the best way to compare them is to plot the Pareto curves.

**Comment by jbloomAus** [comment]
Very excited about this!

File: sae_lens/training/sparse_autoencoder.py [hunk 1]
```diff
@@ -403,6 +404,194 @@ def calculate_ghost_grad_loss(
...
+        hidden_pre = noisy_hidden_pre
+        del noisy_hidden_pre, sae_in
+
+        # gated sae stuff
+        hidden_pre_mag = hidden_pre * torch.exp(self.r_mag) + self.b_mag
+        hidden_post_mag = self.hook_hidden_pre(self.activation_fn(hidden_pre_mag))
+        del hidden_pre_mag
+        # 1 if positive, 0 if negative
+        hidden_pre_gate = hidden_pre + self.b_gate
+        hidden_post_gate = (torch.sign(hidden_pre_gate) + 1) / 2
```
Comment (dtch1997) [inline sae_lens/training/sparse_autoencoder.py:521]: Could be implemented as `hidden_pre_gate > 0`

File: sae_lens/training/sparse_autoencoder.py [hunk 2]
```diff
@@ -403,6 +404,194 @@ def calculate_ghost_grad_loss(
...
+
+        mse_loss = per_item_mse_loss.mean()
+        # l1 loss only applied to pre-gate magnitude
+        # NOTE: in paper, they use ReLU for this
+        # Unsure if we should hardcode ReLU also
+        via_gate_feature_magnitudes = F.relu(hidden_pre_gate)
+        sparsity = via_gate_feature_magnitudes.norm(p=self.lp_norm, dim=1).mean(
+            dim=(0,)
+        )
+        l1_loss = self.l1_coefficient * sparsity
```
[the remark this answers is not in the document]
Comment (dtch1997) [reply sae_lens/training/sparse_autoencoder.py:566]: NOTE: Actually, L0 is calculated based on `feature_acts`, see `train_sae_on_language_model.py:479` so this is fine!

**Comment by dtch1997** [comment]
Report from initial training runs: 

https://wandb.ai/sae-experiments/sae-experiments/reports/Gated-SAE-analysis--Vmlldzo3ODE5OTAw?accessToken=wcmc1fc13wlz3dze7zl26cmkblfdt6984fi20n79fgaa4x79l4moygyclx31mx3q

**Comment by dtch1997** [comment]
Final WandB report: https://wandb.ai/sae-experiments/sae-experiments/reports/Gated-SAE-analysis--Vmlldzo3ODE5OTAw?accessToken=wcmc1fc13wlz3dze7zl26cmkblfdt6984fi20n79fgaa4x79l4moygyclx31mx3q

Pareto plot: 
image


TODO: 
- Clean up the cluster files
- Merge latest changes from main

**Comment by ArthurConmy** [comment]
Thanks for the work so far! I read through the Gated SAE loss func implementation and it seemed right. If setting off runs is cheap, it would be great to get more Gated and Baseline runs in the L0 range 10-100 -- that's the most important region and there are not too many points there

**Comment by jbloomAus** [comment]
Thanks for this! I just merged a seperate PR for this but will look into your transcoder PR shortly!
